src/librustc_typeck/collect.rs

```diff
@@ -1340,23 +1347,106 @@ fn find_existential_constraints<'a, 'tcx>(
+                // skip binder is ok, since we only use this to find generic parameters and their
+                // positions.
+                for (idx, subst) in substs.iter().enumerate() {
+                    if let UnpackedKind::Type(ty) = subst.unpack() {
+                        if let ty::Param(p) = ty.sty {
+                            if index_map.insert(p, idx).is_some() {
+                                // there was already an entry for `p`, meaning a generic parameter
+                                // was used twice
+                                self.tcx.sess.span_err(
+                                    span,
+                                    &format!("defining existential type use restricts existential \
+                                    type by using the generic parameter `{}` twice", p.name),
+                                );
+                                return;
+                            }
+                        } else {
+                            self.tcx.sess.delay_span_bug(
+                                span,
+                                &format!(
+                                    "non-defining exist ty use in defining scope: {:?}, {:?}",
+                                    concrete_type, substs,
+                                ),
+                            );
+                        }
+                    }
+                }
+                // compute the index within the existential type for each generic parameter used in
+                // the concrete type
+                let indices = concrete_type
+                    .subst(self.tcx, substs)
+                    .walk()
+                    .filter_map(|t| match &t.sty {
+                        ty::Param(p) => Some(*index_map.get(p).unwrap()),
+                        _ => None,
+                    }).collect();
+                let is_param = |ty: ty::Ty| match ty.sty {
+                    ty::Param(_) => true,
+                    _ => false,
+                };
+                if !substs.types().all(is_param) {
+                    self.tcx.sess.span_err(
+                        span,
+                        "defining existential type use does not fully define existential type",
+                    );
+                } else if let Some((prev_span, prev_ty, ref prev_indices)) = self.found {
+                    let mut ty = concrete_type.walk().fuse();
+                    let mut p_ty = prev_ty.walk().fuse();
+                    let iter_eq = (&mut ty).zip(&mut p_ty).all(|(t, p)| match (&t.sty, &p.sty) {
+                        // type parameters are equal to any other type parameter for the purpose of
+                        // concrete type equality, as it is possible to obtain the same type just
+                        // by passing matching parameters to a function.
+                        (ty::Param(_), ty::Param(_)) => true,
+                        _ => t == p,
```

Comment:
I'm suspicious of this entire approach. Only the outermost type could be a differing `ty::Param`s, anything *containing* `ty::Param` would automatically hit `t == p` being `false`.

There are two ways to compare types deeply, and they're:
1. exact equality (which, with the right substitutions, can be reached for types coming from different contexts)
2. unification (`ty::relate` or full-on inference, which uses the former) - this is how e.g. `impl` matching works in the current trait system

"Walking" a type is a *not* a good way to achieve almost anything, the main legitimate usecase I've found is gathering leaves of some kind (e.g. `ty::Infer(_)`) or searching for a specific type.